lib/translator/inquiry.cc

```diff
@@ -152,6 +152,66 @@ void TranslateUnitSerialNumberVpd(
   WriteValue(result, buffer);
   WriteValue(product_serial_number, buffer.subspan(sizeof(result)));
 }
+
+void TranslateBlockLimitsVpd(const nvme::IdentifyControllerData& identify_ctrl,
+                             absl::Span<uint8_t> buffer) {
+  // The value is in units of the minimum memory
+  // page size (CAP.MPSMIN) and is reported as a power of two (2^n).
+  // A value of 0h indicates that there is no maximum data transfer size
+  uint32_t max_transfer_length;
+  if (identify_ctrl.mdts > 32) {
+    // Max transfer length should be 2^16 according to section 3.13 of
+    // https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+    DebugLog("max transfer length is > 2^32");
+    max_transfer_length = 1 << 16;
+  } else {
+    max_transfer_length = identify_ctrl.mdts ? 1 << identify_ctrl.mdts : 0;
+  }
+
+  // TODO: put in common?
+  const uint8_t kMaxCompareWriteLen = 255;
+  // compare and write length is 8 bits. we ensure the size fits.
+  uint32_t compare_and_write_len = (max_transfer_length > kMaxCompareWriteLen)
+                                       ? kMaxCompareWriteLen
+                                       : max_transfer_length;
```

Comment:
Similarly, we could define max and min functions if you think they'd improve readability. We could add a number of utility functions in a different PR. Or there might already exists some bazel target with common definitions we can use.

Reply:
interesting, will create utilis issue